src/System.Private.CoreLib/src/System/StartupHookProvider.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+using Internal.IO;
+
+namespace System
+{
+    internal class StartupHookProvider
+    {
+        private static string initializeMethodName = "Initialize";
+
+        // Parse a string specifying a list of assemblies and types
+        // containing a startup hook, and call each hook in turn.
+        private static void ProcessStartupHooks(IntPtr startupHooksVariable)
+        {
+            Debug.Assert(startupHooksVariable != IntPtr.Zero);
+            string[] startupHooks = Marshal.PtrToStringUTF8(startupHooksVariable).Split(new char[] { Path.PathSeparator });
+
+            // Parse startup hook variable
+            List<Tuple<string, string>> startupHooksParsed = new List<Tuple<string, string>>();
+            foreach (string startupHook in startupHooks)
+            {
+                string[] assemblyPathAndTypeName = startupHook.Split(new char[] { '!' });
+                if (assemblyPathAndTypeName.Length != 2)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                string assemblyPath = assemblyPathAndTypeName[0];
+                string typeName = assemblyPathAndTypeName[1];
+                if (String.IsNullOrEmpty(assemblyPath) || String.IsNullOrEmpty(typeName))
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                startupHooksParsed.Add(Tuple.Create(assemblyPath, typeName));
+            }
+
+            // Ensure the startup dlls exist
+            foreach (var startupHook in startupHooksParsed)
+            {
+                var assemblyPath = startupHook.Item1;
+                if (!File.Exists(assemblyPath))
```

Comment:
I would skip this upfront check completely. `LoadFromAssemblyPath` should throw exception that should be good enough to figure out what went wrong (note that there are many other situations how it can fail - e.g. if the filename is not a valid managed dll for example).

If `LoadFromAssemblyPath` does not have good diagnostic today, it is a bug we should fix in `LoadFromAssemblyPath`; not by adding a code here.

Reply:
I think it makes sense to report the missing files before any hooks execute. The exception thrown by LoadFromAssemblyPath is fine - I was just trying to improve diagnostics for the startup hook. For example, if someone adds a startup hook like:
```/path/to/loggingAssembly.dll!LogProvider.Initializer;/path/to/missingFile.dll!TypeName```
it would be good to get the missing file error early, without potential spew from the log provider.

Reply:
I do not think that the extra code is worth it for the minor cosmetic improvement.

If the potential spew from the log provider is long or otherwise distracting, it is the log provider problem - you will get it every time the program is invoked.

My 2 cents...